### ISSUE TITLE
Countdown/ETA task behavior

### ISSUE THREAD

**Issue by averwind** [celery/celery#2541]
Assume the following scenario where there is only on worker:

- The prefetch is set to 1
- There are some number of long running tasks queued with countdown set.
- Since all of them are tasks with countdown the worker gets all the tasks as it is the only one.
- Eventually worker starts working on one task
- The other tasks cannot proceed as prefetch count is 1 even if countdown timer has expired for them
- A new worker is started
- Even though there is an available worker the tasks are all kept in the original worker.

Doesn't it make more sense to re-queue these tasks if the local worker is not able to process them so that another can at least fetch the task?

**Comment by minuscorp** [comment]
I'm not totally sure but, you can use the -Ofair option when launching the worker to prevent that behaviour.
The -Ofair option disables the task prefetching, so when a new worker is spawned It should be consuming the tasks that are not being executed.

**Comment by averwind** [comment]
Right but in this case it doesn't help. Queuing a task with countdown automatically increases the prefetch number so that the worker starts a local timer in itself. Once timer expires the prefetch is decremented.

**Comment by minuscorp** [comment]
I'm sorry I couldn't help you more. I also have an application with 
long-term tasks all programmed with a default countdown. But even when 
the worker amount can't increase itself, I'm not having that problem. 
But I suppose that the last spawned worker doesn't consume already 
counting down tasks from another worker.

El 19/03/2015 a las 23:32, averwind escribió:
>
> Right but in this case it doesn't help. Queuing a task with countdown 
> automatically increases the prefetch number so that the worker starts 
> a local timer in itself. Once timer expires the prefetch is decremented.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/celery/celery/issues/2541#issuecomment-83787346>.
>

-- 

--
Jorge Revuelta Herrero
Computer Science Degree

**Comment by averwind** [comment]
Indeed, again to me sounds like if the worker that completed the countdown is not able to run the task it should pass it back to the queue without any countdown so that another worker may pick it up.

**Comment by minuscorp** [comment]
I totally agree with you, mate. I hope some maintainer reads this.

El 19/03/2015 a las 23:44, averwind escribió:
>
> Indeed, again to me sounds like if the worker that completed the 
> countdown is not able to run the task it should pass it back to the 
> queue without any countdown so that another worker may pick it up.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/celery/celery/issues/2541#issuecomment-83789079>.
>

-- 

--
Jorge Revuelta Herrero
Computer Science Degree

**Comment by VladimirPal** [comment]
@ask Can you answer please what happen when worker that completed the countdown is not able to run the task?

**Comment by thedrow** [comment]
I think that Countdown/ETA expresses that you want to run the task in either 30 seconds or not at all.
We can have a flag that says that a task should be enqueued for processing even if the countdown is reached.

**Comment by VladimirPal** [comment]
@thedrow Such behavior can increase productivity.

**Comment by averwind** [comment]
@thedrow so what you mean is once the countdown is complete the task will go back to the queue instead of staying in the worker that did the countdown?

Another idea (well this would require a revamp) can perhaps be utilizing:
http://www.rabbitmq.com/blog/2015/04/16/scheduling-messages-with-rabbitmq/

**Comment by thedrow** [comment]
Yes exactly.

Thae plugin is very interesting.
We can't relay on the plugin being installed.
We might also be able to use it for celery beat.
I'll add it to the Kickstarter ideas at #2655.

**Comment by averwind** [comment]
Right, that being a plugin is a problem. The workaround mentioned in there as well can be another approach.

http://blog.james-carr.org/2012/03/30/rabbitmq-sending-a-message-to-be-consumed-later/

**Comment by ask** [comment]
That method requires one queue per message, so is not really a viable alternative

**Comment by ask** [comment]
Closing this, as we don't have the resources to complete this task.

**Comment by safwanebrahim** [comment]
We are facing same issue, We have long running tasks and there are more than 50 workers to consume the task. We are ready to solve this issue , What is the right approach to fix this ?

**Comment by thedrow** [comment]
@ask ^

**Comment by superprat** [comment]
+1 willing to contribute some time and effort to resolve this issue as well. Anyone have a status update before we get started?

**Comment by thedrow** [comment]
I was thinking that we may want to provide ETA/Countdown backends.
We could use RabbitMQ's delayed messages to implement ETA/Countdown if it's available and enabled.
We could also relay on other broker specific implementations and only fallback to our current method when none are available.
We can change the current implementation to re-queue all messages that should be executed except the first.

That's the status for the moment.

**Comment by meorilehr** [comment]
@auvipy hey, we also encounter this issue, pretty similar to what @safwanebrahim-uc is encountering. are there any updates about this issues' solution?

**Comment by patrys** [comment]
What if Celery offered an option to push tasks with future ETAs to the end of the queue?

The worst-case scenario is that a few delayed tasks get juggled for a bit until one of them is ready to be executed. This still beats having potentially unrestricted memory consumption on the worker side.

Our Celery workers get killed from time to time because of high memory consumption caused by a buildup of retries with backoff (backoff sets countdown/ETA and causes tasks to be queued locally).
